lrlex/src/lib/ctbuilder.rs

```diff
@@ -1183,99 +1183,228 @@ impl CTLexer {
-                _ => &k,
-            };
-            let tok_ident = format_ident!("T_{}", name.to_ascii_uppercase());
-            (
-                quote! {
-                    #tok_ident,
-                },
-                quote! {
-                    pub const #tok_ident: #storaget = #id;
-                },
-            )
-        })
-        .unzip();
-    // Since the formatter doesn't preserve comments and we don't want to lose build time,
-    // just format the module contents.
-    let unformatted = quote! {
-        mod #mod_ident {
-            #![allow(dead_code)]
-            #tokens
-            pub const TOK_IDS: &[#storaget] = &[#token_array];
+/// [custom lexer example]: https://github.com/softdevteam/grmtools/tree/master/lrlex/examples/calc_manual_lex
+#[derive(Debug, Clone)]
+pub struct CTTokenMapBuilder<StorageT: Display + ToTokens> {
+    mod_name: String,
+    token_map: Vec<(String, TokenStream)>,
+    rename_map: Option<HashMap<String, String>>,
+    allow_dead_code: bool,
+    _marker: PhantomData<StorageT>,
+}
+
+impl<StorageT: Display + ToTokens> CTTokenMapBuilder<StorageT> {
+    /// Create a new token map builder.
+    ///
+    /// See the [builder documentation] for more info.
+    ///
+    /// [builder documentation]: CTTokenMapBuilder
+    pub fn new(
+        mod_name: impl Into<String>,
+        token_map: impl Borrow<HashMap<String, StorageT>>,
+    ) -> Self {
+        Self {
+            mod_name: mod_name.into(),
+            token_map: token_map
+                .borrow()
+                .iter()
+                .map(|(tok_name, tok_value)| (tok_name.clone(), tok_value.to_token_stream()))
+                .collect(),
+            rename_map: None,
+            allow_dead_code: false,
+            _marker: PhantomData,
         }
     }
-    .to_string();
-    let out_mod = syn::parse_str(&unformatted)
-        .map(|syntax_tree| prettyplease::unparse(&syntax_tree))
-        .unwrap_or(unformatted);
-    outs.push_str(&out_mod);
-    let mut outp = PathBuf::from(var("OUT_DIR")?);
-    outp.push(mod_name);
-    outp.set_extension("rs");
-
-    // If the file we're about to write out already exists with the same contents, then we
-    // don't overwrite it (since that will force a recompile of the file, and relinking of the
-    // binary etc).
-    if let Ok(curs) = read_to_string(&outp) {
-        if curs == outs {
-            return Ok(());
+
+    /// Set a token rename map.
+    ///
+    /// Rename map is used to specify identifier names for tokens whose names
+    /// are not valid Rust identifiers. For example, if `token_map`
+    /// is `HashMap{"+": 0, "ID": 1}` and `rename_map` is `HashMap{"+": "PLUS"}`
+    /// then the generated module will look roughly as follows:
+    ///
+    /// ```rust,ignore
+    /// mod x {
+    ///   pub const T_PLUS: u8 = 0;
+    ///   pub const T_ID: u8 = 1;
+    /// }
+    /// ```
+    pub fn set_rename_map<M, I, K, V>(&mut self, rename_map: Option<M>)
```

Comment:
Do we need `set_rename_map` and `rename_map`? It seems more idiomatic to me to have just the latter.